src/main/java/macrobase/analysis/classify/EWAppxPercentileOutlierClassifier.java

```diff
@@ -0,0 +1,104 @@
+import macrobase.util.Periodic;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/*
+ Exponentially weighted approximate percentile-based streaming classifier
+ */
+public class EWAppxPercentileOutlierClassifier extends OutlierClassifier {
+    private ExponentiallyBiasedAChao<DatumWithNorm> reservoir;
+
+    private double currentThreshold = 0;
+
+    private final Periodic reservoirDecayer;
+    private final Periodic percentileUpdater;
+
+    List<OutlierClassificationResult> warmupOutput = new ArrayList<>();
+
+    public EWAppxPercentileOutlierClassifier(MacroBaseConf conf, Iterator<Datum> input) {
+        this(conf,
+             input,
+             conf.getInt(MacroBaseConf.SCORE_RESERVOIR_SIZE, MacroBaseDefaults.SCORE_RESERVOIR_SIZE),
+             conf.getInt(MacroBaseConf.WARMUP_COUNT, MacroBaseDefaults.WARMUP_COUNT),
+             conf.getDecayType(),
+             conf.getDouble(MacroBaseConf.SUMMARY_UPDATE_PERIOD, MacroBaseDefaults.SUMMARY_UPDATE_PERIOD),
+             conf.getDecayType(),
+             conf.getDouble(MacroBaseConf.SUMMARY_UPDATE_PERIOD, MacroBaseDefaults.SUMMARY_UPDATE_PERIOD),
+             conf.getDouble(MacroBaseConf.DECAY_RATE, MacroBaseDefaults.DECAY_RATE),
+             conf.getDouble(MacroBaseConf.TARGET_PERCENTILE, MacroBaseDefaults.TARGET_PERCENTILE));
+    }
+
+    private void updateThreshold(double percentile) {
+        List<DatumWithNorm> norms = reservoir.getReservoir();
+        norms.sort((a, b) -> a.getNorm().compareTo(b.getNorm()));
+        currentThreshold = norms.get((int)(percentile * norms.size())).getNorm();
+    }
+
+    public EWAppxPercentileOutlierClassifier(MacroBaseConf conf,
+                                             Iterator<Datum> input,
+                                             int sampleSize,
+                                             int warmupCount,
+                                             MacroBaseConf.PeriodType updatePeriodType,
+                                             double updatePeriod,
+                                             MacroBaseConf.PeriodType decayPeriodType,
+                                             double decayPeriod,
+                                             double decayRate,
+                                             double percentile) {
+        super(conf, input);
+        reservoir = new ExponentiallyBiasedAChao<>(sampleSize, decayRate, conf.getRandom());
+
+        this.percentileUpdater = new Periodic(updatePeriodType,
+                                              updatePeriod,
+                                              () -> updateThreshold(percentile));
+
+        this.reservoirDecayer = new Periodic(decayPeriodType,
+                                             decayPeriod,
+                                             reservoir::advancePeriod);
+
+        List<DatumWithNorm> warmupInput = new ArrayList<>(warmupCount);
+        for(int i = 0; i < warmupCount; ++i) {
```

Comment:
this is general strategy we used. I think we also drain up iterators in constructors in general too. You have a better design suggestion where to handle this? We just end up with cleaner code by doing this rather than checking if it needs to be done from inside `.next()`